cloudferrylib/utils/retrying.py

```diff
@@ -0,0 +1,199 @@
+        self.wait_exponential = wait_exponential
+        self.max_time = max_time
+        self.raise_error = raise_error
+        self.retry_on_return_value = retry_on_return_value
+        self.return_value = return_value
+        self.expected_exceptions = expected_exceptions or []
+        self.predicate = predicate
+        self.predicate_args = predicate_args
+        self.predicate_kwargs = predicate_kwargs
+        self.predicate_retval_as_arg = predicate_retval_as_arg
+        self.retry_message = retry_message
+        self.stop_on_return_value = stop_on_return_value
+
+    def max_attempts_reached(self):
+        return self.attempt >= self.max_attempts
+
+    def increment_attempts(self):
+        self.attempt += 1
+
+    def timedout(self):
+        return self.total_time >= self.max_time
+
+    def update_total_time(self):
+        if self.wait_exponential:
+            self.timeout *= 2
+
+        self.total_time += self.timeout
+
+    def sleep(self):
+        time.sleep(self.timeout)
+
+    def get_retry_message(self, auto_message):
+        if self.retry_message:
+            return self.retry_message
+        else:
+            return auto_message
+
+    def handle_predicate(self, retval):
+        if self.predicate and callable(self.predicate):
+            if self.predicate_retval_as_arg:
+                args = [retval]
+            else:
+                args = self.predicate_args or []
+            kwargs = self.predicate_kwargs or {}
+            if not self.predicate(*args, **kwargs):
+                default_msg = "Function exited with '{}'".format(retval)
+                msg = self.get_retry_message(default_msg)
+                raise RetryRequired(msg)
+
+    def handle_return_on_value(self, retval):
+        if self.stop_on_return_value and retval != self.return_value:
+            msg = "Return value '{}' didn't match expected '{}'"
+            raise RetryRequired(msg.format(retval, self.return_value))
+
+        if self.retry_on_return_value and retval == self.return_value:
+            msg = self.get_retry_message(
+                "Unexpected return value: {}".format(retval))
+            raise RetryRequired(msg)
+
+    def run(self, func, *args, **kwargs):
+        stop_retrying = self.max_attempts_reached
+        step_action = self.increment_attempts
+
+        if self.max_time is not UNLIMITED:
+            stop_retrying = self.timedout
+            step_action = self.update_total_time
+
+        retval = None
+        failing = True
+        last_exception = None
+
+        while failing and not stop_retrying():
+            step_action()
+            try:
+                retval = func(*args, **kwargs)
+
+                self.handle_predicate(retval)
+                self.handle_return_on_value(retval)
+
+                failing = False
+            except Exception as e:
+                if any([isinstance(e, expected)
+                        for expected in self.expected_exceptions]):
```

Comment:
First, you don't need list comprehension here, you can write `any(isinstance(e, expected) for expected in self.expected_exceptions)`. This will create generator instead of list. Second, if you will be using tuple for `self.expected_exceptions`, then you can write `is instance(e, self.expected_exceptions)`

Consider using tuple to store expected exceptions, like self.expected_exceptions = tuple(expected_exceptions or []), this way you will be able to use isinstance without loop, e.g. isinstance(obj, self.expected_exceptions